nifi-nar-bundles/nifi-deltalake-bundle/nifi-deltalake-processors/src/main/java/org/apache/nifi/processors/deltalake/DeltaLakeMetadataWriter.java

```diff
@@ -0,0 +1,264 @@
+            .required(true)
+            .build();
+
+    public static final PropertyDescriptor GCP_BUCKET = new PropertyDescriptor.Builder()
+            .name("gcp-bucket-url")
+            .displayName("GCP bucket url")
+            .description("The GCP bucket url, has to starts with gs://")
+            .dependsOn(STORAGE_SELECTOR, GCP)
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+            .required(true)
+            .build();
+
+    public static final PropertyDescriptor GCP_PATH = new PropertyDescriptor.Builder()
+            .name("gcp-data-path")
+            .displayName("Data path in GCP bucket")
+            .description("The path to the directory containing the parquet files within the GCP bucket")
+            .dependsOn(STORAGE_SELECTOR, GCP)
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+            .required(true)
+            .build();
+
+    public static final PropertyDescriptor STRUCTURE_JSON = new PropertyDescriptor.Builder()
+            .name("parquet-structure")
+            .displayName("Parquet structure in json")
+            .description("Describes the data structure of the parquet file in json format")
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+            .required(true)
+            .build();
+
+    public static final Relationship REL_SUCCESS = new Relationship.Builder()
+            .name("success")
+            .description("DeltaLake table successfully updated")
+            .build();
+
+    public static final Relationship REL_FAILED = new Relationship.Builder()
+            .name("failure")
+            .description("DeltaLake table update failed")
+            .build();
+
+    private DeltaLakeService deltalakeService;
+
+    @Override
+    protected void init(final ProcessorInitializationContext context) {
+        this.relationships = Collections.unmodifiableSet(new HashSet<>(Arrays.asList(REL_SUCCESS, REL_FAILED)));
+        this.descriptors = Collections.unmodifiableList(Arrays.asList(STORAGE_SELECTOR, LOCAL_PATH, STRUCTURE_JSON,
+                S3_ACCESS_KEY, S3_SECRET_KEY, S3_BUCKET, S3_PATH,
+                AZURE_ACCOUNT_KEY, AZURE_STORAGE_NAME, AZURE_STORAGE_ACCOUNT, AZURE_PATH,
+                GCP_ACCOUNT_JSON_KEYFILE_PATH, GCP_BUCKET, GCP_PATH));
+        deltalakeService = new DeltaLakeService();
+    }
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return this.relationships;
+    }
+
+    @Override
+    public final List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return descriptors;
+    }
+
+    @Override
+    public void onTrigger(final ProcessContext context, final ProcessSession session) throws ProcessException {
+        FlowFile flowFile = session.create();
+        try {
+            Map<String, String> updateResult = updateDeltaLake();
+            session.putAllAttributes(flowFile, updateResult);
```

Comment:
Please transfer the updated flowfile:

FlowFile updatedFlowFile = session.putAllAttributes(flowFile, updateResult);
session.transfer(updatedFlowFile, REL_FAILED);

The same applies for session.putAttribute() usage.


Note: REL_FAILURE would be a more consistent name (since the name of the relationship is "failure").

Reply:
Transfered and renamad, thanks.